[SIL] Witness methods store the conformance from which they come.

huonw opened the pull request.

huonw committed:
[SIL] Witness methods store the conformance from which they come.

huonw commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -2213,12 +2225,26 @@ namespace {
-                                  getASTContext());
+      Optional<ProtocolConformanceRef> witnessMethodConformance;
+      if (auto conformance = origType->getWitnessMethodConformanceOrNone()) {
+        if (conformance->isAbstract()) {
```
@slavapestov I don't understand what substitution I'm meant to being here to convert the abstract conformance into a concrete one (and even for the concrete case, I'm just guessing).

huonw commented:
@swift-ci please smoke test

huonw committed:
[test] Update SIL printing/parsing tests for 'witness_method: <protocol>'.

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please smoke test

huonw commented:
@swift-ci please smoke test.

huonw commented:
@swift-ci please smoke test.

huonw commented:
@swift-ci please smoke test.

huonw commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -2213,12 +2225,27 @@ namespace {
+      if (auto conformance = origType->getWitnessMethodConformanceOrNone()) {
+        assert(origType->getExtInfo().hasSelfParam());
+        auto selfType = origType->getSelfParameter().getType();
+        // The Self type can be nested in a few layers of metatypes (etc.), e.g.
```
Example:

```swift
protocol Foo {
  static var x: Int { get set }
}

func mutate(_: inout Int) {}
func foo<T: Foo>(ty: T.Type) {
  mutate(&T.x)
}

```

The SIL for the setter callback call after calling materializeForSet looks like:

```
...
bb2(%16 : $Builtin.RawPointer):                   // Preds: bb0
  %17 = pointer_to_thin_function %16 : $Builtin.RawPointer to $@convention(witness_method: Foo) <τ_0_0 where τ_0_0 : Foo> (Builtin.RawPointer, @inout Builtin.UnsafeValueBuffer, @inout @thick τ_0_0.Type, @thick τ_0_0.Type.Type) -> () // user: %22
...
```

huonw commented:
@swift-ci please smoke test and merge.

huonw commented:
@swift-ci please smoke test and merge.

huonw commented:
@swift-ci please smoke test and merge.

huonw commented:
@swift-ci please smoke test and merge.

huonw commented:
@swift-ci please smoke test and merge.

slavapestov commented:
Are you planning on doing serialization in a separate commit? If so LGTM!

slavapestov commented on include/swift/AST/Types.h:
```diff
@@ -3547,6 +3550,20 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
+  /// If this is a @convention(witness_method) function, return the conformance
+  /// for which the method is a witness, if it isn't that convention, return
+  /// None.
+  Optional<ProtocolConformanceRef> getWitnessMethodConformanceOrNone() const {
```
Is this variant useful?

huonw replied:
It's the one I call most often: it avoids having do a `fnTy->getRepresentation() == SILFunctionRepresentation::WitnessMethod ? fnTy.getWitnessMethodConformance() : None` dance.

slavapestov replied:
I see, OK.

huonw commented:
Serialization is in the first commit.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test.

huonw commented:
Failure on linux is:

```
Stack dump:
0.	Program arguments: /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swift-linux-x86_64/bin/swift -frontend -c /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/DependencyResolver.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraph.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraphLoader.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/PackageGraphRoot.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/RawPackageConstraints.swift /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/PackageGraph/RepositoryPackageContainerProvider.swift -target x86_64-unknown-linux -disable-objc-interop -sdk / -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/clibc/include -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/foundation-linux-x86_64/Foundation -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/foundation-linux-x86_64/Foundation/usr/lib/swift -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/xctest-linux-x86_64 -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/libdispatch-linux-x86_64/src -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/libdispatch-linux-x86_64/src/swift -I /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swift-corelibs-libdispatch -enable-testing -module-cache-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/ModuleCache -swift-version 4 -O -D SWIFT_PACKAGE -Xcc -fmodule-map-file=/home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/swiftpm/Sources/clibc/include/module.modulemap -Xcc -fblocks -emit-module-doc-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.swiftdoc -parse-as-library -module-name PackageGraph -emit-module-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.swiftmodule -emit-dependencies-path /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraph.d -num-threads 32 -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/DependencyResolver.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraph.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraphLoader.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/PackageGraphRoot.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/RawPackageConstraints.swift.o -o /home/buildnode/jenkins/workspace/swift-PR-Linux-smoke-test@2/branch-master/buildbot_linux/swiftpm-linux-x86_64/x86_64-unknown-linux/release/PackageGraph.build/RepositoryPackageContainerProvider.swift.o 
1.	While running pass #61987 SILFunctionTransform "Indirect Call Devirtualization" on SILFunction "@_T0s8SequencePss9Equatable7ElementRpzrlE8containsSbAEFSay12PackageGraph22ResolvedProductBuilder33_04246CEF6CEAB47798DA707D73AF3E55LLCG_Tg5".
 for 'contains(_:)' in module 'Swift'
<unknown>:0: error: unable to execute command: Aborted
```

huonw commented:
@swift-ci Please smoke test.

huonw commented:
@swift-ci Please smoke test.

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test.

Maybe this will work some day?!

shahmishal commented:
@swift-ci smoke test

huonw committed:
[AST] Handle classes properly in ProtocolConformanceRef::subst.

Previously ProtocolConformance::subst would crash because it was receiving
things with an unexpected relationship between the conformance's type and the
substituted self type. The compiler doesn't quite properly model "abstract"
inherited conformances, so we end up using normal conformances instead, and we
need to work around this in some cases.

huonw commented:
@swift-ci smoke test.

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test

huonw commented:
@swift-ci smoke test and merge

rudkx commented:
@swift-ci Please smoke test

shahmishal commented:
@swift-ci Please smoke test Linux

shahmishal commented:
@swift-ci Please smoke test Linux

huonw commented:
@swift-ci smoke test and merge

huonw commented:
@swift-ci smoke test and merge

huonw commented:
@swift-ci Please smoke test and merge

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test

huonw commented:
@swift-ci Please smoke test
